File: src/aoc2015/jav/Day17.java
package aoc2015.jav;

import java.io.IOException;
import java.nio.file.*;
import java.util.*;
import java.util.stream.Collectors;

class Day17 {

  private static List<Integer> partOne(List<String> list) {
    List<Integer> containers = list.stream().mapToInt(Integer::valueOf).boxed().collect(Collectors.toList());
    List<Integer> solutionLength = new ArrayList<>();
    System.out.println("Part One = " + solve(150, 0, containers, solutionLength));
    return solutionLength;
  }

  private static long partTwo(List<Integer> solutions) {
    int min = solutions.stream().min(Integer::compareTo).get();
    return solutions.stream().filter(x -> x.equals(min)).count();
  }

  private static int solve(int liters, int used, List<Integer> unused, List<Integer> sLength) {
    if (liters == 0) {
      sLength.add(used);
      return 1;
    } else if (liters < 0 || unused.isEmpty())
      return 0;
    else {
      List<Integer> tail = unused.subList(1, unused.size());
      return solve(liters - unused.get(0), used + 1, tail, sLength) + solve(liters, used, tail, sLength);
    }
  }

  public static void main(String[] args) throws IOException {
    List<String> s = Files.readAllLines(Paths.get("./input/2015/Day17_input.txt"));
    List<Integer> solutions = partOne(s);
    System.out.println("Part Two = " + partTwo(solutions));
  }
}

File: src/aoc2017/kot/Day17.kt
package aoc2017.kot

object Day17 {
  fun partOne(stepSize: Int, steps: Int): Int {
    var pos = 0
    val lock = mutableListOf(0)

    repeat(steps) {
      pos = (pos + stepSize) % lock.size + 1
      lock.add(pos, lock.size)
    }
    return lock[pos + 1]
  }

  fun partTwo(stepSize: Int, steps: Int): Int {
    var pos = 0
    var valueAfterZero = -1

    for (size in 1..steps) {
      pos = (pos + stepSize) % size + 1
      if (pos == 1) valueAfterZero = size
    }
    return valueAfterZero
  }
}

fun main(args: Array<String>) {
  println("Part One = ${Day17.partOne(304, 2017)}")
  println("Part Two = ${Day17.partTwo(304, 50_000_000)}")
}
